### ISSUE TITLE
Events not consumed by bloc on Widget tests

### ISSUE THREAD

**Issue by geovin89** [felangel/bloc#1349]
**Describe the bug**
I have created the login form exactly like your tutorial Firebase Login tutorial (it worked fine) and wrote a widget test to unit test it

```dart
class MockLoginRepository extends Mock implements LoginRepository  {}

void main() {
  LoginRepository loginRepository;
  LoginInBloc loginInBloc;

  setUpAll(() {
    //configureInjections(_injectable.Environment.test);
    loginRepository = MockLoginRepository();
    loginInBloc = loginInBloc(loginRepository);
  });

testWidgets(
        'should have sign in button enabled after entering email/password',
        (WidgetTester tester) async {
      await tester.pumpWidget(
        BlocProvider.value(
          value: loginInBloc,
          child: const MaterialApp(
            home: Scaffold(body: LogInForm()),
          ),
        ),
      );

      final loginButtonFinder = find.byKey(Key(UIKeys.kLogInButtonKey));
      expect(loginButtonFinder, findsOneWidget);
      expect(
        (loginButtonFinder.evaluate().first.widget as MaterialButton).enabled,
        false,
      );

      final emailFieldFinder = find.byKey(Key(UIKeys.kEmailFieldKey));
      await tester.enterText(emailFieldFinder, 'geo@gmail.com');
      await tester.pumpAndSettle();

      final passwordFieldFinder = find.byKey(Key(UIKeys.kPasswordFieldKey));
      await tester.enterText(passwordFieldFinder, 'Geo@123');
      await tester.pumpAndSettle();

      expect(
        (loginButtonFinder.evaluate().first.widget as MaterialButton).enabled,
        true,
      );
    });
}
```
the test fails.

**Expected behavior**
The loginButton should get enabled and the test should pass

**Logs**
I added logs to LoginForm class to check whether the functions are called
Logs: 
```
called _isSignInButtonEnabled
called _onLoginInEmailChanged with email 
called _onLoginEmailChanged with email geo@gmail.com
called _onLoginPasswordChanged with pwd
called _onLoginPasswordChanged with pwd Geo@123
```

I added print statements in mapState function in loginBloc too but none of them are getting printed, which leads me to believe that the bloc is not consuming the events. Can you please help me identify what I am doing wrong?

**Comment by felangel** [comment]
Hi @geovin89 👋 
Thanks for opening an issue!

I would highly recommend mocking the bloc in your widget test instead of using a real bloc with a mock repository. Ideally, you should mock everything except for the thing you're testing which in this case is the `LoginForm`.

```dart
class MockLoginBloc extends MockBloc<LoginEvent, LoginState> implements LoginBloc {}

void main() {
  LoginBloc loginBloc;

  setUp(() {
    loginBloc = MockLoginBloc();
  });

  testWidgets('sign in button should be enabled when bloc state is enabled', (tester) async {
    when(loginBloc.state).thenReturn(LoginState(valid: true, ...);
    await tester.pumpWidget(BlocProvider.value(value: loginBloc, child: MaterialApp(home: Scaffold(body: LoginForm())));
    expect(tester.widget<MaterialButton>(find.byKey(UIKeys.kLogInButtonKey)).enabled, isTrue);
  });
}
```

Hope that helps 👍 

Closing for now but feel free to comment if you have any additional questions. If after trying this approach you are still having trouble it would be great if you could share a link to a sample repo which reproduces the issue so I can take a closer look 😄

**Comment by geovin89** [comment]
According to your code, the login button only gets enabled when there is a value in the email field and the password field and the other thing that I have found is 'when(loginBloc.state).thenReturn ' doesn't work when you are testing something in the blocListner code.

I have created a sample app here https://github.com/geovin89/flutterbloctestapp. I have posted my questions there as 3 Todos to make your life easier.

Thanks for taking the time
